New re-used repo dataset/repo instances need different invalidation logic

Here is the demo

````
>>> from datalad.api import create
>>> ds = create('here')
>>> ds.remove()
>>> ds = create('here')
[INFO] Creating a new annex repo at /tmp/bugdemo/here 
[ERROR] Failed to run ['git', 'config', '--file', '/tmp/bugdemo/here/.datalad/config', '--unset-all', 'datalad.dataset.id'] under '/tmp/bugdemo/here'. Exit code=5. out= err= 
````

Reason is: in `create` it unsets config variables if they are present. The config of the removed dataset is still loaded in the cfg manager of that dataset, but there is nothing in the actual config file, hence `git config` fails.

--- a/datalad/distribution/tests/test_uninstall.py
+++ b/datalad/distribution/tests/test_uninstall.py
@@ -21,6 +21,7 @@ from datalad.api import uninstall
 from datalad.api import drop
 from datalad.api import remove
 from datalad.api import install
+from datalad.api import create
 from datalad.support.exceptions import InsufficientArgumentsError, CommandError
 from datalad.tests.utils import ok_
 from datalad.tests.utils import eq_
@@ -347,3 +348,17 @@ def test_kill(path):
     assert_raises(CommandError, ds.remove)
     eq_(ds.remove(recursive=True, check=False), [subds, ds])
     ok_(not exists(path))
+
+
+@with_tempfile()
+def test_remove_recreation(path):
+
+    # test recreation is possible and doesn't conflict with in-memory
+    # remainings of the old instances
+    # see issue #1311
+
+    ds = create(path)
+    ds.remove()
+    ds = create(path)
+    ok_clean_git(ds.path)
+    ok_(ds.is_installed())

--- a/datalad/distribution/uninstall.py
+++ b/datalad/distribution/uninstall.py
@@ -76,6 +76,8 @@ def _uninstall_dataset(ds, check, has_super):
     if has_super and not exists(ds.path):
         # recreate an empty mountpoint to make Git happier
         os.makedirs(ds.path)
+    # invalidate loaded ConfigManager:
+    ds._cfg = None
     results.append(ds)
     return results
 

